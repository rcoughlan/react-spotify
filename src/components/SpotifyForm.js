import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
    Button,
    Col,
    Form,
    Row,
    ToggleButtonGroup,
    ToggleButton
} from 'react-bootstrap';

class SpotifyForm extends Component {


    handleChange = (e) => {
        const updateFields = this.props.updateFields;
        e.preventDefault();
        let obj = {};
        obj[e.target.name] = e.target.value;
        updateFields(obj);
    }

    onSubmit = (e) => {
        e.preventDefault();
        this.props.hitApi();
    }

    render() {
        return (

            <Row className="search-form">
                <Col></Col>
                <Col sm="10" md="8" lg="6" className="search">

                    <Form onSubmit={this.onSubmit}>

                        <Form.Group >
                            <ToggleButtonGroup name="searchType" defaultValue="top-tracks" className="mt-3 toggle-search">
                                <ToggleButton id="top-tracks" variant="success" type="radio" value="top-tracks" onChange={this.handleChange}>Top Tracks</ToggleButton>
                                <ToggleButton id="top-artists" variant="success" type="radio" value="top-artists" onChange={this.handleChange}>Top Artists</ToggleButton>
                                <ToggleButton id="recent-tracks" variant="success" type="radio" value="recent-tracks" onChange={this.handleChange}>Recent Tracks</ToggleButton>
                            </ToggleButtonGroup>
                        </Form.Group>
                        {
                            this.props.searchType === 'recent-tracks' ?
                                <Form.Row>
                                    <Col></Col>
                                        <Form.Group as={Col}>
                                            <Form.Label>Limit</Form.Label>
                                            <Form.Control id="limit" name="limit" as="select" onChange={this.handleChange}>
                                                <option value={false}>none</option>
                                                <option>10</option>
                                                <option>20</option>
                                                <option>30</option>
                                                <option>40</option>
                                                <option>50</option>
                                            </Form.Control>
                                        </Form.Group>
                                    <Col></Col>
                                </Form.Row>
                                :
                                <Form.Row>
                                    <Form.Group as={Col}>
                                        <Form.Label>Time Range</Form.Label>
                                        <Form.Control id="time-range" className="time-range" name="timeRange" as="select" onChange={this.handleChange}>
                                            <option id="none" value={false}>none</option>
                                            <option id="short_term" value={"short_term"}>short term</option>
                                            <option id="medium_term" value={"medium_term"}>medium term</option>
                                            <option id="long_term" value={"long_term"}>long term</option>
                                        </Form.Control>
                                    </Form.Group>
                                    <Form.Group as={Col}>
                                        <Form.Label>Offset</Form.Label>
                                        <Form.Control id="offset" name="offset" as="select" onChange={this.handleChange}>
                                            <option value={false}>none</option>
                                            <option>10</option>
                                            <option>20</option>
                                            <option>30</option>
                                            <option>40</option>
                                            <option>50</option>
                                        </Form.Control>
                                    </Form.Group>
                                    <Form.Group as={Col}>
                                        <Form.Label>Limit</Form.Label>
                                        <Form.Control id="limit" name="limit" as="select" onChange={this.handleChange}>
                                            <option value={false}>none</option>
                                            <option>10</option>
                                            <option>20</option>
                                            <option>30</option>
                                            <option>40</option>
                                            <option>50</option>
                                        </Form.Control>
                                    </Form.Group>
                                </Form.Row>
                        }

                        <Button variant="success" type="submit" className="submit">Submit</Button>
                    </Form>
                </Col>
                <Col></Col>
            </Row>
        );
    }
}

SpotifyForm.propTypes = {
    hitApi: PropTypes.func.isRequired,
    updateFields: PropTypes.func.isRequired,
    searchType: PropTypes.string
}

export default SpotifyForm;