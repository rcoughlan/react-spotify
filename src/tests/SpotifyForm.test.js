import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import SpotifyForm from '../components/SpotifyForm'

describe('SpotifyForm Component', () => {
    let state = [];
    const hitApi = jest.fn(null);

    it('matches the snapshot', () => {
        const handleUpdateFields = jest.fn(x => state.push(x));
        const tree = renderer.create(<SpotifyForm updateFields={handleUpdateFields} hitApi={hitApi} />).toJSON();

        expect(tree).toMatchSnapshot();
    })

    it('has 3 toggle buttons', () => {
        const handleUpdateFields = jest.fn(x => state.push(x));
        const wrapper = shallow(<SpotifyForm updateFields={handleUpdateFields} hitApi={hitApi} />);
        const toggleGroup = wrapper.find('.toggle-search');
        expect(toggleGroup.children()).toHaveLength(3);
    });
 
    it('submit button calls hitApi()', () => {
        const handleUpdateFields = jest.fn(x => state.push(x));
        const wrapper = mount(<SpotifyForm updateFields={handleUpdateFields} hitApi={hitApi} />);
        const form = wrapper.find('form');
        form.simulate('submit');
        expect(hitApi).toHaveBeenCalledTimes(1);
    });
    
    it('has toggle buttons for search types with one default active', () => {
        const handleUpdateFields = jest.fn(x => state.push(x));
        const wrapper = mount(<SpotifyForm updateFields={handleUpdateFields} hitApi={hitApi} />);
        const topTracks = wrapper.find('label#top-tracks');
        const topArtists = wrapper.find('label#top-artists');
        const topArtistsInput = topArtists.childAt(0);
        expect(topTracks.hasClass('active')).toBe(true);
        expect(topArtists.hasClass('active')).toBe(false);
        expect(topArtistsInput.prop('type')).toEqual('radio');
    });
    
    it('changes the value of the limit dropdown', () => {
        const handleUpdateFields = jest.fn(x => state.push(x));
        const wrapper = mount(<SpotifyForm updateFields={handleUpdateFields} hitApi={hitApi} />);
        const limit = wrapper.find('select#limit');
        expect(limit).toHaveLength(1);
        limit.simulate('change', { value: ['20'] });
        expect(handleUpdateFields).toBeCalledTimes(1);
    });

    it('changes the value of the time range dropdown', () => {
        const handleUpdateFields = jest.fn(x => state.push(x));
        const wrapper = mount(<SpotifyForm updateFields={handleUpdateFields} hitApi={hitApi} />);
        const timeRange = wrapper.find('select#time-range');
        expect(timeRange).toHaveLength(1);
        timeRange.simulate('change', { value: ['short_term'] });
        expect(handleUpdateFields).toBeCalledTimes(1);
    });

    it('changes the value of the offset dropdown', () => {
        const handleUpdateFields = jest.fn(x => state.push(x));
        const wrapper = mount(<SpotifyForm updateFields={handleUpdateFields} hitApi={hitApi} />);
        const offset = wrapper.find('select#offset');
        expect(offset).toHaveLength(1);
        offset.simulate('change', { value: 10 });
        expect(handleUpdateFields).toBeCalledTimes(1);
    });
})