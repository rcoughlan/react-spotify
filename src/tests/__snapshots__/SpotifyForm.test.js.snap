// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpotifyForm Component matches the snapshot 1`] = `
<div
  className="row"
>
  <div
    className="col"
  />
  <div
    className="search col-lg-6 col-md-8 col-sm-10"
  >
    <form
      className=""
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          className="auth-input form-label"
        >
          Enter your auth token
        </label>
        <input
          className="auth-input form-control"
          name="auth"
          onChange={[Function]}
          placeholder="Auth Token"
          type="text"
        />
        <small
          className="form-text"
        >
          This can be found at spotify.api.com
        </small>
      </div>
      <div
        className="form-group"
      >
        <div
          className="mt-3 toggle-search btn-group btn-group-toggle"
          role="group"
        >
          <label
            className="btn active btn-success"
            disabled={false}
            id="top-tracks"
            type={null}
          >
            <input
              autoComplete="off"
              checked={true}
              disabled={false}
              name="searchType"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              type="radio"
              value="top-tracks"
            />
            Top Tracks
          </label>
          <label
            className="btn btn-success"
            disabled={false}
            id="top-artists"
            type={null}
          >
            <input
              autoComplete="off"
              checked={false}
              disabled={false}
              name="searchType"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              type="radio"
              value="top-artists"
            />
            Top Artists
          </label>
          <label
            className="btn btn-success"
            disabled={false}
            id="recent-tracks"
            type={null}
          >
            <input
              autoComplete="off"
              checked={false}
              disabled={false}
              name="searchType"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              type="radio"
              value="recent-tracks"
            />
            Recent Tracks
          </label>
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          className="form-group col"
        >
          <label
            className="form-label"
          >
            Time Range
          </label>
          <select
            className="time-range form-control"
            id="time-range"
            name="timeRange"
            onChange={[Function]}
          >
            <option
              id="none"
              value={false}
            >
              none
            </option>
            <option
              id="short_term"
              value="short_term"
            >
              short term
            </option>
            <option
              id="medium_term"
              value="medium_term"
            >
              medium term
            </option>
            <option
              id="long_term"
              value="long_term"
            >
              long term
            </option>
          </select>
        </div>
        <div
          className="form-group col"
        >
          <label
            className="form-label"
          >
            Offset
          </label>
          <select
            className="form-control"
            id="offset"
            name="offset"
            onChange={[Function]}
          >
            <option
              value={false}
            >
              none
            </option>
            <option>
              10
            </option>
            <option>
              20
            </option>
            <option>
              30
            </option>
            <option>
              40
            </option>
            <option>
              50
            </option>
          </select>
        </div>
        <div
          className="form-group col"
        >
          <label
            className="form-label"
          >
            Limit
          </label>
          <select
            className="form-control"
            id="limit"
            name="limit"
            onChange={[Function]}
          >
            <option
              value={false}
            >
              none
            </option>
            <option>
              10
            </option>
            <option>
              20
            </option>
            <option>
              30
            </option>
            <option>
              40
            </option>
            <option>
              50
            </option>
          </select>
        </div>
      </div>
      <button
        className="submit btn btn-success"
        disabled={false}
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
  <div
    className="col"
  />
</div>
`;
